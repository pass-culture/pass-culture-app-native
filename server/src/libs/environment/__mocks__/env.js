"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.env = void 0;
exports.env = {
    ENV: 'test',
    APP_PUBLIC_URL: 'http://localhost:8080',
    APP_PROXY_URL: 'https://app-proxy.testing.passculture.team',
    API_BASE_URL: 'https://api.testing.passculture.team',
    API_BASE_PATH_NATIVE_V1: 'native/v1',
    DEEPLINK_PROTOCOL: 'passculture://',
    PROXY_CACHE_CONTROL: 'public,max-age=3600',
};
