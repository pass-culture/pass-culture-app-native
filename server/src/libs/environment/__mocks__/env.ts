import { Environment } from '../types'

export const env: Environment = {
  ENV: 'test',
  APP_PUBLIC_URL: 'http://localhost:8080',
  APP_PROXY_URL: 'https://app-proxy.testing.passculture.team',
  API_BASE_URL: 'https://api.testing.passculture.team',
  API_BASE_PATH_NATIVE_V1: 'native/v1',
  DEEPLINK_PROTOCOL: 'passculture://',
  PROXY_CACHE_CONTROL: 'public,max-age=3600',
}
