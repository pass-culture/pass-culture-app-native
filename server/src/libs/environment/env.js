"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.env = void 0;
const path_1 = require("path");
const dotenv_1 = require("dotenv");
(0, dotenv_1.config)({
    path: (0, path_1.resolve)(__dirname, '../../..', `.env.${process.env.ENV}`),
});
exports.env = process.env;
