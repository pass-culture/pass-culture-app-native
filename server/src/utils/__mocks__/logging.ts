/* eslint-disable */

export const logger = {
  info(...args: any) {
    jest.fn(...args)
  },
  warn(...args: any) {
    jest.fn(...args)
  },
  error(...args: any) {
    jest.fn(...args)
  },
}
