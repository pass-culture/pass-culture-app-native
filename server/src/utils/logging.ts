export { logger } from 'firebase-functions'
