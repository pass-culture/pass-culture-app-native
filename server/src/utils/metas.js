"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceHtmlMetas = void 0;
const env_1 = require("../libs/environment/env");
const apiClient_1 = require("../services/apiClient");
const types_1 = require("../services/entities/types");
const { APP_PUBLIC_URL } = env_1.env;
const { href } = new URL(APP_PUBLIC_URL);
function replaceHtmlMetas(html, endpoint, type, entityId) {
    return __awaiter(this, void 0, void 0, function* () {
        const entity = yield (0, apiClient_1.apiClient)(type, entityId);
        const subPath = endpoint.slice(1);
        const METAS_CONFIG = types_1.ENTITY_METAS_CONFIG_MAP[type];
        const metaConfig = {
            title: {
                data: METAS_CONFIG.title(entity),
                regEx: /<meta\s(name)="(title)"\scontent="([^"]*)"\s*\/?>/g,
            },
            description: {
                data: METAS_CONFIG.description(entity),
                regEx: /<meta\s(name)="(description)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['og:url']: {
                data: METAS_CONFIG['og:url'](entity, href, subPath),
                regEx: /<meta\s(property)="(og:url)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['og:title']: {
                data: METAS_CONFIG['og:title'](entity),
                regEx: /<meta\s(property)="(og:title)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['og:description']: {
                data: METAS_CONFIG['og:description'](entity),
                regEx: /<meta\s(property)="(og:description)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['og:image']: {
                data: METAS_CONFIG['og:image'](entity),
                regEx: /<meta\s(property)="(og:image)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['og:image:alt']: {
                data: METAS_CONFIG['og:image:alt'](entity),
                regEx: /<meta\s(property)="(og:image:alt)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:card']: {
                data: METAS_CONFIG['twitter:card'](),
                regEx: /<meta\s(name)="(twitter:card)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:url']: {
                data: METAS_CONFIG['twitter:url'](entity, href, subPath),
                regEx: /<meta\s(name)="(twitter:url)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:title']: {
                data: METAS_CONFIG['twitter:title'](entity),
                regEx: /<meta\s(name)="(twitter:title)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:description']: {
                data: METAS_CONFIG['twitter:description'](entity),
                regEx: /<meta\s(name)="(twitter:description)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:image']: {
                data: METAS_CONFIG['twitter:image'](entity),
                regEx: /<meta\s(name)="(twitter:image)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:image:alt']: {
                data: METAS_CONFIG['twitter:image:alt'](entity),
                regEx: /<meta\s(name)="(twitter:image:alt)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:app:url:iphone']: {
                data: METAS_CONFIG['twitter:app:url:iphone'](entity, href, subPath),
                regEx: /<meta\s(name)="(twitter:app:url:iphone)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:app:url:ipad']: {
                data: METAS_CONFIG['twitter:app:url:ipad'](entity, href, subPath),
                regEx: /<meta\s(name)="(twitter:app:url:ipad)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['twitter:app:url:googleplay']: {
                data: METAS_CONFIG['twitter:app:url:googleplay'](entity, href, subPath),
                regEx: /<meta\s(name)="(twitter:app:url:googleplay)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['al:ios:url']: {
                data: METAS_CONFIG['al:ios:url'](entity, href, subPath),
                regEx: /<meta\s(name)="(al:ios:url)"\scontent="([^"]*)"\s*\/?>/g,
            },
            ['al:android:url']: {
                data: METAS_CONFIG['al:android:url'](entity, href, subPath),
                regEx: /<meta\s(name)="(al:android:url)"\scontent="([^"]*)"\s*\/?>/g,
            },
        };
        Object.values(metaConfig).forEach(({ regEx, data }) => {
            if (data) {
                html = html.replace(regEx, `<meta $1="$2" content="${data}" />`);
            }
        });
        return html;
    });
}
exports.replaceHtmlMetas = replaceHtmlMetas;
