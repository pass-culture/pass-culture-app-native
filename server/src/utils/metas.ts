import { encode } from 'html-entities'

import { env } from '../libs/environment/env'
import { apiClient } from '../services/apiClient'
import { ENTITY_METAS_CONFIG_MAP, EntityKeys, MetaConfig } from '../services/entities/types'

import { logger } from './logging'

const { APP_PUBLIC_URL, ORGANIZATION_PREFIX } = env

const { href } = new URL(APP_PUBLIC_URL)

const REGEX = {
  metaTitle: /<meta\s+(name)="(title)"\s+content="([^"]*)"\s*\/?>/g,
  metaDescription: /<meta\s+(name)="(description)"\s+content="([^"]*)"\s*\/?>/g,
  ['og:url']: /<meta\s+(property)="(og:url)"\s+content="([^"]*)"\s*\/?>/g,
  ['og:title']: /<meta\s+(property)="(og:title)"\s+content="([^"]*)"\s*\/?>/g,
  ['og:description']: /<meta\s+(property)="(og:description)"\s+content="([^"]*)"\s*\/?>/g,
  ['og:image']: /<meta\s+(property)="(og:image)"\s+content="([^"]*)"\s*\/?>/g,
  ['og:image:alt']: /<meta\s+(property)="(og:image:alt)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:card']: /<meta\s+(name)="(twitter:card)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:url']: /<meta\s+(name)="(twitter:url)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:title']: /<meta\s+(name)="(twitter:title)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:description']: /<meta\s+(name)="(twitter:description)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:image']: /<meta\s+(name)="(twitter:image)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:image:alt']: /<meta\s+(name)="(twitter:image:alt)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:app:url:iphone']:
    /<meta\s+(name)="(twitter:app:url:iphone)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:app:url:ipad']: /<meta\s+(name)="(twitter:app:url:ipad)"\s+content="([^"]*)"\s*\/?>/g,
  ['twitter:app:url:googleplay']:
    /<meta\s+(name)="(twitter:app:url:googleplay)"\s+content="([^"]*)"\s*\/?>/g,
  ['al:ios:url']: /<meta\s+(name)="(al:ios:url)"\s+content="([^"]*)"\s*\/?>/g,
  ['al:android:url']: /<meta\s+(name)="(al:android:url)"\s+content="([^"]*)"\s*\/?>/g,
}

export function addOrganizationSuffix(title: string) {
  return `${title} | ${ORGANIZATION_PREFIX}`
}

export async function replaceHtmlMetas(
  html: string,
  endpoint: string,
  type: EntityKeys,
  entityId: number
) {
  try {
    const entity = await apiClient(type, entityId)
    const subPath = endpoint.slice(1)
    const METAS_CONFIG = ENTITY_METAS_CONFIG_MAP[type]
    const metaConfig: MetaConfig = {
      metaTitle: {
        data: addOrganizationSuffix(METAS_CONFIG.metaTitle(entity)),
        regEx: REGEX.metaTitle,
      },
      metaDescription: {
        data: METAS_CONFIG.metaDescription(entity),
        regEx: REGEX.metaDescription,
      },
      ['og:url']: {
        data: METAS_CONFIG['og:url'](href, subPath),
        regEx: REGEX['og:url'],
      },
      ['og:title']: {
        data: addOrganizationSuffix(METAS_CONFIG['og:title'](entity)),
        regEx: REGEX['og:title'],
      },
      ['og:description']: {
        data: METAS_CONFIG['og:description'](entity),
        regEx: REGEX['og:description'],
      },
      ['og:image']: {
        data: METAS_CONFIG['og:image'](entity),
        regEx: REGEX['og:image'],
      },
      ['og:image:alt']: {
        data: METAS_CONFIG['og:image:alt'](entity),
        regEx: REGEX['og:image:alt'],
      },
      ['twitter:card']: {
        data: METAS_CONFIG['twitter:card'](),
        regEx: REGEX['twitter:card'],
      },
      ['twitter:url']: {
        data: METAS_CONFIG['twitter:url'](entity, href, subPath),
        regEx: REGEX['twitter:url'],
      },
      ['twitter:title']: {
        data: addOrganizationSuffix(METAS_CONFIG['twitter:title'](entity)),
        regEx: REGEX['twitter:title'],
      },
      ['twitter:description']: {
        data: METAS_CONFIG['twitter:description'](entity),
        regEx: REGEX['twitter:description'],
      },
      ['twitter:image']: {
        data: METAS_CONFIG['twitter:image'](entity),
        regEx: REGEX['twitter:image'],
      },
      ['twitter:image:alt']: {
        data: METAS_CONFIG['twitter:image:alt'](entity),
        regEx: REGEX['twitter:image:alt'],
      },
      ['twitter:app:url:iphone']: {
        data: METAS_CONFIG['twitter:app:url:iphone'](entity, href, subPath),
        regEx: REGEX['twitter:app:url:iphone'],
      },
      ['twitter:app:url:ipad']: {
        data: METAS_CONFIG['twitter:app:url:ipad'](entity, href, subPath),
        regEx: REGEX['twitter:app:url:ipad'],
      },
      ['twitter:app:url:googleplay']: {
        data: METAS_CONFIG['twitter:app:url:googleplay'](entity, href, subPath),
        regEx: REGEX['twitter:app:url:googleplay'],
      },
      ['al:ios:url']: {
        data: METAS_CONFIG['al:ios:url'](entity, href, subPath),
        regEx: REGEX['al:ios:url'],
      },
      ['al:android:url']: {
        data: METAS_CONFIG['al:android:url'](entity, href, subPath),
        regEx: REGEX['al:android:url'],
      },
    }

    html = html.replace(
      /<title>.*<\/title>/g,
      `<title>${addOrganizationSuffix(encode(METAS_CONFIG.title(entity)))}</title>`
    )

    Object.values(metaConfig).forEach(({ regEx, data }) => {
      if (data) {
        html = html.replace(regEx, `<meta $1="$2" content="${encode(data)}" />`)
      }
    })

    if (entity['metadata']) {
      const metadata = recursiveEncode(entity['metadata'])

      html = html.replace(
        '</head>',
        `  <script type="application/ld+json">${JSON.stringify(metadata)}</script>\n  </head>`
      )
    }

    return html
  } catch (error) {
    logger.error(error)
    return html
  }
}

type NestedMetadata = string | number | boolean | string[] | { [key: string]: NestedMetadata }
type Metadata = Record<string, NestedMetadata>

const encodeWhateverType = (value: NestedMetadata): NestedMetadata => {
  switch (typeof value) {
    case 'object':
      if (Array.isArray(value)) {
        return value.map((item) => encode(item))
      }
      return recursiveEncode(value)
    case 'string':
      return encode(value)
    default:
      return value
  }
}

const recursiveEncode = (entityMetadata: Metadata): Metadata => {
  return entityMetadata
    ? Object.entries(entityMetadata)
        .map<[string, NestedMetadata]>(([key, value]) => [key, encodeWhateverType(value)])
        .reduce((newMetadata, [key, value]) => ({ ...newMetadata, [key]: value }), {})
    : {}
}
