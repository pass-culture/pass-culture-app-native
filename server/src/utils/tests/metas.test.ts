import {
  OFFER_RESPONSE_SNAPSHOT,
  OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA,
  OFFER_RESPONSE_SNAPSHOT_WITH_VENUE_NULL_METADATA,
  TEST_HTML,
  VENUE_WITH_BANNER_RESPONSE_SNAPSHOT,
} from '../../../tests/constants'
import { ENTITY_MAP, EntityKeys } from '../../services/entities/types'
import { replaceHtmlMetas, addOrganizationPrefix } from '../metas'
import { env } from '../../libs/environment/env'
import { logger } from '../logging'

const mockEntityMap = ENTITY_MAP
const mockOfferResponse = OFFER_RESPONSE_SNAPSHOT
const mockOfferResponseWithDangerousMetadata = OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA
const mockOfferResponseWithVenueNullMetadata = OFFER_RESPONSE_SNAPSHOT_WITH_VENUE_NULL_METADATA
const mockVenueResponse = VENUE_WITH_BANNER_RESPONSE_SNAPSHOT
const INTERNAL_SERVER_ERROR_STATUS_CODE = 502

jest.mock('../logging', () => jest.requireMock('../__mocks__/logging'))

jest.mock('../../services/apiClient', () => ({
  apiClient: async (type: EntityKeys, id: number) => {
    if (mockEntityMap[type].API_MODEL_NAME === 'offer' && id !== INTERNAL_SERVER_ERROR_STATUS_CODE) {
      if (id === mockOfferResponseWithDangerousMetadata.id) {
        return mockOfferResponseWithDangerousMetadata
      }
      if (id === mockOfferResponseWithVenueNullMetadata.id) {
        return mockOfferResponseWithVenueNullMetadata
      }
      return mockOfferResponse
    } else if (mockEntityMap[type].API_MODEL_NAME === 'venue') {
      return mockVenueResponse
    }
    throw new Error(`Wrong status code: ${INTERNAL_SERVER_ERROR_STATUS_CODE}`)
  },
}))

describe('metas utils', () => {
  it(`should replace meta for offer`, async () => {
    const newHtml = await replaceHtmlMetas(
      TEST_HTML,
      `/offre/${OFFER_RESPONSE_SNAPSHOT.id}`,
      'offre' as EntityKeys,
      OFFER_RESPONSE_SNAPSHOT.id
    )
    expect(newHtml).toMatchSnapshot()
  })

  it(`should replace meta for venue`, async () => {
    const newHtml = await replaceHtmlMetas(
      TEST_HTML,
      `/lieu/${VENUE_WITH_BANNER_RESPONSE_SNAPSHOT.id}`,
      'lieu' as EntityKeys,
      VENUE_WITH_BANNER_RESPONSE_SNAPSHOT.id
    )
    expect(newHtml).toMatchSnapshot()
  })

  it(`should prefix string "Hello World" with "${env.ORGANIZATION_PREFIX} |"`, () => {
    expect(addOrganizationPrefix('Hello World')).toEqual(`${env.ORGANIZATION_PREFIX} | Hello World`)
  })

  it('should encode metadata in json-ld', async () => {
    const html = await replaceHtmlMetas(
      TEST_HTML,
      `/offre/${OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id}`,
      'offre' as EntityKeys,
      OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id
    )

    expect(html).toContain(
      '&lt;/script&gt;&lt;script&gt;alert(&quot;you have been pwned&quot;)&lt;/script&gt;'
    )
  })

  it('should encode metadata in json-ld recursively', async () => {
    const html = await replaceHtmlMetas(
      TEST_HTML,
      `/offre/${OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id}`,
      'offre' as EntityKeys,
      OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id
    )

    expect(html).toContain(
      '&lt;/script&gt;&lt;script&gt;alert(&quot;you have been hacked&quot;)&lt;/script&gt;'
    )
  })

  it('should encode metadata in meta tags', async () => {
    const html = await replaceHtmlMetas(
      TEST_HTML,
      `/offre/${OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id}`,
      'offre' as EntityKeys,
      OFFER_RESPONSE_SNAPSHOT_WITH_DANGEROUS_METADATA.id
    )

    expect(html).toContain(
      '&lt;/script&gt;&lt;script&gt;alert(&quot;you have been pranked&quot;)&lt;/script&gt;'
    )
  })

  it('should not break when venue meta have null values', async () => {
    await expect(replaceHtmlMetas(
      TEST_HTML,
      `/offre/${OFFER_RESPONSE_SNAPSHOT_WITH_VENUE_NULL_METADATA.id}`,
      'offre' as EntityKeys,
      OFFER_RESPONSE_SNAPSHOT_WITH_VENUE_NULL_METADATA.id
    )).resolves.not.toThrow()
  })

  it(`should return basic html in case of apiClient throwing ${INTERNAL_SERVER_ERROR_STATUS_CODE} status code`, async () => {
    const html = await replaceHtmlMetas(
      TEST_HTML,
      `/offre/${INTERNAL_SERVER_ERROR_STATUS_CODE}`,
      'offre' as EntityKeys,
      INTERNAL_SERVER_ERROR_STATUS_CODE
    )

    expect(html).toEqual(TEST_HTML)
    expect(logger.error).toBeCalledWith(new Error(`Wrong status code: 502`))
  })
})
