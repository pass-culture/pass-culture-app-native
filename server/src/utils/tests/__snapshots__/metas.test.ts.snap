// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metas utils should replace meta for offer 1`] = `
"<!doctype html>
<html lang=\\"fr\\">
  <head>
    <meta charset=\\"utf-8\\">
    <title>&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF | pass Culture</title>
    <meta name=\\"title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF | pass Culture\\" />
    <meta name=\\"description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Depuis de nombreuses années, Christine vit sous un pont, isolée de toute famille et amis. Par une nuit comme il n’en existe que dans les contes, un jeune garçon de 8 ans fait irruption devant son abri. Suli ne parle pas français, il est perdu, séparé de sa mère…
Tous les détails du film sur AlloCiné: https://www.allocine.fr/film/fichefilm_gen_cfilm=199293.html\\" />
    <meta name=\\"author\\" content=\\"pass Culture\\"/>
    <meta name=\\"keywords\\" content=\\"CINE PLEIN AIR\\" />
    <meta property=\\"og:type\\" content=\\"website\\"/>
    <meta property=\\"og:url\\" content=\\"http://localhost:8080/offre/116656\\" />
    <meta property=\\"og:title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF | pass Culture\\" />
    <meta property=\\"og:description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Depuis de nombreuses années, Christine vit sous un pont, isolée de toute famille et amis. Par une nuit comme il n’en existe que dans les contes, un jeune garçon de 8 ans fait irruption devant son abri. Suli ne parle pas français, il est perdu, séparé de sa mère…
Tous les détails du film sur AlloCiné: https://www.allocine.fr/film/fichefilm_gen_cfilm=199293.html\\" />
    <meta property=\\"og:image\\" content=\\"https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/products/CHSYS\\" />
    <meta property=\\"og:image:alt\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF\\" />
    <meta property=\\"og:locale\\" content=\\"fr_FR\\"/>
    <meta property=\\"og:site_name\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:card\\" content=\\"summary\\" />
    <meta name=\\"twitter:url\\" content=\\"http://localhost:8080/offre/116656\\" />
    <meta name=\\"twitter:title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF | pass Culture\\" />
    <meta name=\\"twitter:description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Depuis de nombreuses années, Christine vit sous un pont, isolée de toute famille et amis. Par une nuit comme il n’en existe que dans les contes, un jeune garçon de 8 ans fait irruption devant son abri. Suli ne parle pas français, il est perdu, séparé de sa mère…
Tous les détails du film sur AlloCiné: https://www.allocine.fr/film/fichefilm_gen_cfilm=199293.html\\" />
    <meta name=\\"twitter:image\\" content=\\"https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/products/CHSYS\\" />
    <meta name=\\"twitter:image:alt\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Sous les étoiles de Paris - VF\\" />
    <meta name=\\"twitter:site\\" content=\\"@pass_Culture\\"/>
    <meta name=\\"twitter:app:country\\" content=\\"FR\\"/>
    <meta name=\\"twitter:app:name:iphone\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:iphone\\" content=\\"1557887412\\"/>
    <meta name=\\"twitter:app:url:iphone\\" content=\\"passculture://http://localhost:8080/offre/116656\\" />
    <meta name=\\"twitter:app:name:ipad\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:ipad\\" content=\\"1557887412\\"/>
    <meta name=\\"twitter:app:url:ipad\\" content=\\"passculture://http://localhost:8080/offre/116656\\" />
    <meta name=\\"twitter:app:name:googleplay\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:googleplay\\" content=\\"app.passculture.testing\\"/>
    <meta name=\\"twitter:app:url:googleplay\\" content=\\"passculture://http://localhost:8080/offre/116656\\" />
    <meta property=\\"al:ios:app_name\\" content=\\"pass Culture\\"/>
    <meta property=\\"al:ios:app_store_id\\" content=\\"1557887412\\"/>
    <meta property=\\"al:ios:url\\" content=\\"passculture://https://app.testing.passculture.team\\"/>
    <meta property=\\"al:android:url\\" content=\\"passculture://https://app.testing.passculture.team\\">
    <meta property=\\"al:android:app_name\\" content=\\"pass Culture\\"/>
    <meta property=\\"al:android:package\\" content=\\"app.passculture.testing\\"/>
    <script type=\\"application/ld+json\\">{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Event\\",\\"name\\":\\"Sous les étoiles de Paris - VF\\",\\"location\\":{\\"@type\\":\\"Place\\",\\"name\\":\\"PATHE BEAUGRENELLE\\",\\"address\\":{\\"@type\\":\\"PostalAddress\\",\\"streetAddress\\":\\"2 RUE LAMENNAIS\\",\\"postalCode\\":\\"75008\\",\\"addressLocality\\":\\"PARIS 8\\"},\\"geo\\":{\\"@type\\":\\"GeoCoordinates\\",\\"latitude\\":\\"20\\",\\"longitude\\":\\"2\\"}}}</script>
  </head>
  <body>
    <p>Hello world</p>
  </body>
</html>"
`;

exports[`metas utils should replace meta for venue 1`] = `
"<!doctype html>
<html lang=\\"fr\\">
  <head>
    <meta charset=\\"utf-8\\">
    <title>&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Le Petit Rintintin 1 | pass Culture</title>
    <meta name=\\"title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Le Petit Rintintin 1 | pass Culture\\" />
    <meta name=\\"description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;https://pass.culture.fr/ lorem ipsum consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, labore nesciunt numquam. Id itaque in sed sapiente blanditiis necessitatibus. consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus,\\" />
    <meta name=\\"author\\" content=\\"pass Culture\\"/>
    <meta name=\\"keywords\\" content=\\"pass Culture\\" />
    <meta property=\\"og:type\\" content=\\"website\\"/>
    <meta property=\\"og:url\\" content=\\"http://localhost:8080/lieu/5543\\" />
    <meta property=\\"og:title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Le Petit Rintintin 1 | pass Culture\\" />
    <meta property=\\"og:description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;https://pass.culture.fr/ lorem ipsum consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, labore nesciunt numquam. Id itaque in sed sapiente blanditiis necessitatibus. consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus,\\" />
    <meta property=\\"og:image\\" content=\\"https://app.testing.passculture.team/images/app-icon-512px.png\\"/>
    <meta property=\\"og:image:alt\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;https://pass.culture.fr/ lorem ipsum consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, labore nesciunt numquam. Id itaque in sed sapiente blanditiis necessitatibus. consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus,\\" />
    <meta property=\\"og:locale\\" content=\\"fr_FR\\"/>
    <meta property=\\"og:site_name\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:card\\" content=\\"summary_large_image\\" />
    <meta name=\\"twitter:url\\" content=\\"http://localhost:8080/lieu/5543\\" />
    <meta name=\\"twitter:title\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;Le Petit Rintintin 1 | pass Culture\\" />
    <meta name=\\"twitter:description\\" content=\\"&lt;img src=&quot;data:image/gif;base64 onerror=&quot;javascript:alert(&quot;hack&quot;)&quot; /&gt;https://pass.culture.fr/ lorem ipsum consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, labore nesciunt numquam. Id itaque in sed sapiente blanditiis necessitatibus. consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus, consectetur adipisicing elit. Debitis officiis maiores quia unde, hic quisquam odit ea quo ipsam possimus,\\" />
    <meta name=\\"twitter:image\\" content=\\"https://app.testing.passculture.team/images/app-icon-512px.png\\"/>
    <meta name=\\"twitter:image:alt\\" content=\\"https://image.png\\" />
    <meta name=\\"twitter:site\\" content=\\"@pass_Culture\\"/>
    <meta name=\\"twitter:app:country\\" content=\\"FR\\"/>
    <meta name=\\"twitter:app:name:iphone\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:iphone\\" content=\\"1557887412\\"/>
    <meta name=\\"twitter:app:url:iphone\\" content=\\"passculture://http://localhost:8080/lieu/5543\\" />
    <meta name=\\"twitter:app:name:ipad\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:ipad\\" content=\\"1557887412\\"/>
    <meta name=\\"twitter:app:url:ipad\\" content=\\"passculture://http://localhost:8080/lieu/5543\\" />
    <meta name=\\"twitter:app:name:googleplay\\" content=\\"pass Culture\\"/>
    <meta name=\\"twitter:app:id:googleplay\\" content=\\"app.passculture.testing\\"/>
    <meta name=\\"twitter:app:url:googleplay\\" content=\\"passculture://http://localhost:8080/lieu/5543\\" />
    <meta property=\\"al:ios:app_name\\" content=\\"pass Culture\\"/>
    <meta property=\\"al:ios:app_store_id\\" content=\\"1557887412\\"/>
    <meta property=\\"al:ios:url\\" content=\\"passculture://https://app.testing.passculture.team\\"/>
    <meta property=\\"al:android:url\\" content=\\"passculture://https://app.testing.passculture.team\\">
    <meta property=\\"al:android:app_name\\" content=\\"pass Culture\\"/>
    <meta property=\\"al:android:package\\" content=\\"app.passculture.testing\\"/>
  </head>
  <body>
    <p>Hello world</p>
  </body>
</html>"
`;
